src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,1045 @@
+
+                # Errors
+
+                When `x` is greater than the maximum storable value in `self`.
+                ```
+                use bit_fields::{BitRangeMut, CheckedAssignError};
+                let mut x = 18", stringify!($x), ";
+
+                let mut nibble = BitRangeMut::<_,0,4>(&mut x);
+                assert_eq!(nibble.write(16),Err(CheckedAssignError));
+                assert_eq!(nibble.write(15),Ok(()));
+
+                let mut nibble = BitRangeMut::<_,4,8>(&mut x);
+                assert_eq!(nibble.write(16),Err(CheckedAssignError));
+                assert_eq!(nibble.write(15),Ok(()));
+                ```
+            ")]
+            #[inline]
+            pub fn write(&mut self, x: $x) -> Result<(), $crate::CheckedAssignError> {
+                self.checked_assign(x)
+            }
+
+            #[doc = concat!("
+                Adds `x` to the value of the bit range.
+
+                # Errors
+
+                1. When `x` is greater than the maximum value storable in the bit range.
+                2. When adding `x` to the value of the bit range would overflow.
+
+                ```
+                use bit_fields::{BitRangeMut, CheckedAddAssignError};
+                let mut x = 18", stringify!($x), ";
+
+                let mut nibble = BitRangeMut::<_,0,4>(&mut x);
+                assert_eq!(nibble,2);
+                assert_eq!(nibble.checked_add_assign(16),Err(CheckedAddAssignError::OutOfRange));
+                assert_eq!(nibble.checked_add_assign(14),Err(CheckedAddAssignError::Overflow));
+                assert_eq!(nibble.checked_add_assign(2),Ok(()));
+                assert_eq!(nibble,4);
+                assert_eq!(x,20);
+                
+                let mut nibble = BitRangeMut::<_,4,8>(&mut x);
+                assert_eq!(nibble,1);
+                assert_eq!(nibble.checked_add_assign(16),Err(CheckedAddAssignError::OutOfRange));
+                assert_eq!(nibble.checked_add_assign(15),Err(CheckedAddAssignError::Overflow));
+                assert_eq!(nibble.checked_add_assign(2),Ok(()));
+                assert_eq!(nibble,3);
+                assert_eq!(x,52);
+                ```
+            ")]
+            #[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+            #[inline]
+            pub fn checked_add_assign(
+                &mut self,
+                x: $x,
+            ) -> Result<(), $crate::CheckedAddAssignError> {
+                if x <= Self::MAX {
+                    let cur = (Self::MASK & *self.0) >> START;
+                    if x <= Self::MAX - cur {
+                        let shift = x << START;
+                        *self.0 += shift;
+                        Ok(())
+                    } else {
+                        Err($crate::CheckedAddAssignError::Overflow)
+                    }
+                } else {
+                    Err($crate::CheckedAddAssignError::OutOfRange)
+                }
```

Comment:
How do you feel about reusing `checked_assign` here, e.g. something like

```suggestion
                self.checked_assign(self.read() + x)
```

Reply:
As discussed offline, we will keep the conditions but re-use `unchecked_assign`.